Concept: Ports (#158)
microsoft/vcpkg-docs: vcpkg/concepts/ports.md

---
# concepts/ports
title: Ports
description: Ports as a vcpkg concept
author: JavierMatosD
ms.author: javiermat
ms.date: 09/21/2023
ms.prod: vcpkg
ms.topic: concept-article
---

# Concept: Ports

> In the context of the vcpkg package management system, a port is a predefined set of instructions and scripts that facilitate the easy integration of a specific library or package into C++ projects. A port comprises a collection of files that dictate how a library is built and installed, thus acting as a blueprint for managing library dependencies in a C++ environment.

Reviewer: ```suggestion
A port is vcpkg's method of delivering packages, at the very least a port contains:

* Metadata about the package: name, version, supported features, dependencies, etc.
* Instructions to acquire, build if necessary, and install the package. 

Additionally, ports often contain:

* Patches to apply to the package's source code
* Instructions to integrate with your project's build system
```

The introductory paragraph feels a bit redundant; I suggest we use a more direct description of what ports are.

> ## Portfile.cmake

Reviewer: ```suggestion
## Portfile (`portfile.cmake`)
```

> The `portfile.cmake` is a script that contains a series of instructions defining how to build and install a library within the vcpkg environment. It outlines the steps to retrieve the library's source code, the methods to build it, and the guidelines to install it, ensuring a smooth integration process.

Reviewer: ```suggestion
All ports must contain a portfile (a file named `portfile.cmake`). A portfile is a script that contains instructions on how to build and install a package within the vcpkg environment. Some common tasks performed by portfiles are: 

* Download a package's source code.
* Build the package using the upstream build system.
* Copy the build output to the vcpkg installation tree.
* Install build system integration files.
```

## vcpkg.json

> The `vcpkg.json` is a metadata file that provides detailed information about the library in question. This file contains various attributes including the library’s name, version, description, and a list of dependencies, which help in streamlining the installation process and ensuring compatibility with other library components.

Reviewer: ```suggestion
The `vcpkg.json` file describes metadata about the package contained in the vcpkg port. This file defines attributes like the package’s name, version, description, dependencies, supported features and platforms, etc.

See the [`vcpkg.json`](../reference/vcpkg-json.md) file reference to learn more.
```

> ## Patches

Reviewer: ```suggestion
## Patch files
```

> In certain scenarios, a port might include patch files. These files carry modifications to the library's source code, which are essential to either fix existing issues or to maintain compatibility with other components. During the build process, these patches are applied automatically, aiding in a seamless and error-free library integration.

Reviewer: ```suggestion
In certain scenarios, a port might include patch files. These files carry modifications to the library's source code, which are essential to either fix existing issues or to maintain compatibility with other components. Before the build process, patches are applied as part of the portfile execution steps, aiding in a seamless and error-free library integration.
```

---

> This concise documentation serves as a guideline for understanding the basic constituents of a vcpkg port, highlighting the core concepts that play a pivotal role in managing and integrating libraries effectively in C++ projects through the vcpkg system.

Reviewer: Replace this with a "Next steps" header

---
title: Ports
description: Ports as a vcpkg concept
author: JavierMatosD
ms.author: javiermat
ms.date: 09/21/2023
ms.prod: vcpkg
ms.topic: concept-article
---

# Concept: Ports

> A port in vcpkg is a versioned recipe that produces a set of files. It also imposes constraints on the install graph, ensuring compatible dependency resolution. At a minimum, a port contains:

Reviewer: ```suggestion
A port in vcpkg is a versioned recipe that produces a set of files. Executing a port affects the installation graph by producing new header or binary files. Ports may have dependencies on other ports and may require additional dependencies when certain features are enabled, these dependencies must be reflected in the ports metadata. At a minimum, a port contains:
```
The original sentence is not clear to me, are we trying to say that ports install things? that ports may depend on other ports? or both?

* Metadata about the package: name, version, supported features, dependencies, etc.
* Instructions to acquire, build if necessary, and install the package.

Additionally, ports may include:

* Patches to apply to the package's source code
* Instructions to integrate with your project's build system

## Portfile (`portfile.cmake`)

All ports must contain a portfile (a file named `portfile.cmake`). A portfile is a script that contains instructions on how to build and install a package within the vcpkg environment. Some common tasks performed by portfiles are:

* Download a package's source code.
* Build the package using the upstream build system.
* Copy the build output to the vcpkg installation tree.
* Install build system integration files.

## vcpkg.json

The `vcpkg.json` file describes metadata about the package contained in the vcpkg port. This file defines attributes like the package’s name, version, description, dependencies, supported features and platforms, etc.
See the [`vcpkg.json`](../reference/vcpkg-json.md) file reference to learn more.

## Patch files

> In certain scenarios, a port might include patch files. These files carry modifications to the library's source code, which are essential to either fix existing issues or to maintain compatibility with other components. Before the build process, patches are applied as part of the portfile execution steps, aiding in a seamless and error-free library integration.

Reviewer 2: Should this link to the part of the maintainer guide?

Reviewer: That's a good suggestion

> ## Port categories

Reviewer 2: I think it would be good to add a paragraph here saying something that port categories aren't really a 'product feature'; they are something maintainers say colloquially to each other but there's no reason that a port can't be several or none of these.

This is kind of blurring the line between documentation and maintainer notes; I think it's OK here though

### Meta port

A meta port serves to impose constraints on the install graph without having its own build files. These ports are often used to group a set of related libraries under a single name or to serve as a layer of indirection for replaceable components. For example, `boost` is a meta-port that groups together various Boost libraries under a single name for easier dependency management.

### Script port

A script port contains utility scripts that are primarily intended for internal tasks within vcpkg or for consumption by other ports. These ports enable versioning of the utility scripts, thereby maintaining compatibility and standardization. For example, `vcpkg-cmake` is a script port that exposes  the `vcpkg_cmake_configure()`
helper function.

## Next steps

This document provides an overview of ports in vcpkg, detailing their essential elements and various categories. For more advanced topics and detailed guides, refer to the following:

> * [Script ports](..\maintainers\authoring-script-ports)

Reviewer: ```suggestion
* [Script ports](..\maintainers\authoring-script-ports.md)
```

* Metadata about the package: name, version, supported features, dependencies, etc.
* Instructions to acquire, build if necessary, and install the package.

Additionally, ports may include:

* Patches to apply to the package's source code
* Instructions to integrate with your project's build system

## Portfile (`portfile.cmake`)

All ports must contain a portfile (a file named `portfile.cmake`). A portfile is a script that contains instructions on how to build and install a package within the vcpkg environment. Some common tasks performed by portfiles are:

* Download a package's source code.
* Build the package using the upstream build system.
* Copy the build output to the vcpkg installation tree.
* Install build system integration files.

## vcpkg.json

The `vcpkg.json` file describes metadata about the package contained in the vcpkg port. This file defines attributes like the package’s name, version, description, dependencies, supported features and platforms, etc.
See the [`vcpkg.json`](../reference/vcpkg-json.md) file reference to learn more.

> ## CONTROL files

Reviewer: ```suggestion
## CONTROL files

> [!WARNING]
> `CONTROL` files are deprecated and only retained for backwards compatibility.
```

Reviewer: LGTM, just requesting a warning for visibility.

## Patch files

In certain scenarios, a port might include patch files. These files carry modifications to the library's source code, which are essential to either fix existing issues or to maintain compatibility with other components. Before the build process, patches are applied as part of the portfile execution steps, aiding in a seamless and error-free library integration.

## Port categories

### Standard port

A standard port serves as a recipe for building libraries from source or providing header files for header-only libraries. It contains the instructions and metadata needed to acquire, build, and install the library.

### Meta port

> A meta port serves to impose constraints on the install graph without having its own build files. These ports are often used to group a set of related libraries under a single name or to serve as a layer of indirection for replaceable components.

Reviewer: Maybe mention `boost` or `qt` as examples of metaports

---
title: Ports

> description: Ports as a vcpkg concept

Reviewer: ```suggestion
description: This article explains the concept of ports, their capabilities, and their expected contents in detail.
```

## Port categories

Port categories are colloquial terms that maintainers use to articulate the roles different ports play within vcpkg. While these categories help describe general functionalities, it's worth noting that any given port can belong to multiple categories or perhaps none at all, based on its features and use-cases.

### Standard port

A standard port serves as a recipe for building libraries from source or providing header files for header-only libraries. It contains the instructions and metadata needed to acquire, build, and install the library. The vast majority of the libraries in the curated registry are standard ports.

### Meta port

A meta port serves to impose constraints on the install graph without having its own build files. These ports are often used to group a set of related libraries under a single name or to serve as a layer of indirection for replaceable components. For example, `boost` is a meta-port that groups together various Boost libraries under a single name for easier dependency management.

### Script port

> A script port contains utility scripts that are primarily intended for internal tasks within vcpkg or for consumption by other ports. These ports enable versioning of the utility scripts, thereby maintaining compatibility and standardization. For example, `vcpkg-cmake` is a script port that exposes the `vcpkg_cmake_configure()`

Reviewer: ```suggestion
A script port contains utility scripts that are primarily intended for internal tasks within vcpkg or for consumption by other ports. These ports enable versioning of the utility scripts, thereby maintaining compatibility and standardization. For example, `vcpkg-cmake` is a script port that exposes the [`vcpkg_cmake_configure()`](../maintainers/functions/vcpkg_cmake_configure.md)
```
